### ISSUE TITLE
Support dynamic overwrite

### ISSUE THREAD

**Issue by kevinjqliu** [apache/iceberg-python#1287]
### Feature Request / Improvement

Currently `overwrite` consists of a delete + append operation. As an optimization, we can support dynamic overwrite for when an entire partition is replaced. 

Heres an example from @koenvo
https://gist.github.com/koenvo/e23bfab32c7e7810eb52f82c6304fc22

**Comment by Fokko** [comment]
Good one @kevinjqliu, however I think this is a duplicate of https://github.com/apache/iceberg-python/pull/931. There is also a PR https://github.com/apache/iceberg-python/pull/931 that's waiting for me (sorry for that)!

**Comment by koenvo** [comment]
Ah the PR contains quite some similar functionality indeed.

It seems that the PR does a delete+append. If I understand correctly, this could lead to reading incomplete data: in case the delete snapshot is referenced during the read, the data for the overwritten partition(s) is missing.

**Comment by kevinjqliu** [comment]
Thanks @Fokko didnt see that one, I'll close this issue when that PR is merged

> It seems that the PR does a delete+append. If I understand correctly, this could lead to reading incomplete data: in case the delete snapshot is referenced during the read, the data for the overwritten partition(s) is missing.

@koenvo I think the delete+append are done in the context of a transaction. When the transaction is committed, both are either registered together at the same time or rejected at the same time.

**Comment by koenvo** [comment]
I also believe those two snapshots are added in a single transaction. What I mean is that it’s possible to time-travel to the delete snapshot. In that case you are looking at data where the delete is already applied but the append is not.

**Comment by kevinjqliu** [comment]
yes, thats right. This will create 2 snapshots and if you time travel to the first one, you will only see the table with the data deleted.
What if your use case here to time travel to only one of the snapshots?

**Comment by koenvo** [comment]
Ah good question. In our normal process the Iceberg tables are only queried using our own application. The application will always (for now at least) use the latest snapshot. That works fine when those two snapshots are committed in a single transactions. But we also expose the tables in AWS Athena for ad-hoc queries. This makes it possible for people to choose any snapshot, including the delete one.

**Comment by kevinjqliu** [comment]
I see thanks for the explanation. When writing in fast append mode (DELETE+APPEND), it's possible to accidentally time travel to the DELETE snapshot and only see the deleted data. 
Let's get the dynamic overwrite PR in, I think that should be written as 1 snapshot.

**Comment by sundaresanr** [comment]
It would be nice to support multiple partition overwrites in a single transaction => produce single snapshot
